torch/fx/experimental/fuser.py

```diff
@@ -0,0 +1,45 @@
+from torch.fx import (
+    GraphModule,
+)
+from torch.nn.utils.fusion import fuse_conv_bn_eval
+import torch
+
+# turn foo.bar -> ['foo', 'bar']
+def _parent_name(target):
+    r = target.rsplit('.', 1)
+    if len(r) == 1:
+        return '', r[0]
+    else:
+        return r[0], r[1]
+
+# Works for length 2 patterns with 2 modules
+def matches_module_pattern(pattern, node, modules):
+    if len(node.args) == 0:
+        return False
+    nodes = [node.args[0], node]
+    for pattern_node, current_node in zip(pattern, nodes):
+        if current_node.op != 'call_module':
+            return False
+        if not isinstance(current_node, torch.fx.Node):
+            return False
+        if current_node.target not in modules:
+            return False
+        if type(modules[current_node.target]) is not pattern_node:
+            return False
+    return True
+
+
+def fuse(model):
+    patterns = [(torch.nn.Conv2d, torch.nn.BatchNorm2d), (torch.nn.Conv1d, torch.nn.BatchNorm1d), (torch.nn.Conv3d, torch.nn.BatchNorm3d)]
+    for pattern in patterns:
+        for node in model.graph.nodes:
+            modules = dict(model.named_modules())
+            if matches_module_pattern(pattern, node, modules):
+                if len(node.args[0].users) > 1:  # Output of conv is used by other nodes
+                    continue
+                fused_conv = fuse_conv_bn_eval(modules[node.args[0].target], modules[node.target])
+                parent_name, name = _parent_name(node.args[0].target)
+                setattr(dict(model.named_modules())[parent_name], name, fused_conv)
```

Comment:
Instead of mutating Modules, can we simply construct a new dict of qualname -> Module and pass that into the `GraphModule` constructor? https://github.com/pytorch/pytorch/blob/cdc2d2843b2bb7ef1f7192718a24bfabb93b3d36/torch/fx/graph_module.py#L152

Line 36 can be hoisted out of the loop nest here and modified when we find a match. Also, we should make a new copy of the Graph. Then we can just pass that dict and the new graph to the GraphModule constructor

Reply:
Kind of resolved? There are issues with just having a single dict, so the quantized fuser does something similar.